src/analyser.cpp

```diff
@@ -2502,17 +2502,27 @@ void Analyser::AnalyserImpl::analyseModel(const ModelPtr &model)
         }
     }
 
+    // Detmerine whether some variables have been marked as external.
+
+    auto hasExternalVariables = std::any_of(mInternalVariables.begin(), mInternalVariables.end(), [](const auto &iv) {
+        return iv->mIsExternal;
+    });
+
     // Loop over our equations, checking which variables, if any, can be
     // determined using a given equation.
-    // Note: we are actually looping twice. The first time, we don't check for
-    //       NLA systems while we do the second time round. The first time we
-    //       loop, we want to handle all the equations that have one unknown
-    //       variable which is is on its own either on the LHS or RHS of an
-    //       equation. Once we have done that then we can see whether what is
-    //       left can be used in one or several NLA systems.
+    // Note: we loop twice by checking the model with the view of:
+    //        1) getting an ODE system WITHOUT any NLA systems; and then
+    //        2) getting an ODE system WITH one or several NLA systems.
+    //       After those two loops, if we still have some unknown variables and
+    //       they have been marked as external, then we consider them as
+    //       initialised and we go through the two loops one more time. This is
+    //       to account for models that have unknown variables (rendering the
+    //       model invalid) that have been marked as external (rendering the
+    //       model valid).
 
     auto stateIndex = MAX_SIZE_T;
     auto variableIndex = MAX_SIZE_T;
+    auto loopNb = 1;
```

Comment:
No super keen on this style of variable naming, but I knew what it was when I read the code so I guess it works.

Reply:
Happy to change it to something else that we are both happy with.

Reply:
`loopNumber`?

Reply:
Fine by me: b618810f.